### PR TITLE
Reachability: Use flatcontainer to encode data

### PR DESCRIPTION
**Pull request by antiguru**
This changes the reachability logging implementation to use flatcontainers
to replay the logging stream, and on the input to the pre-arranged data.
It choses not to change the input to the final arrangement to limit the
scope of this PR. This serves as a testebed to see what's needed to get
flatcontainers into more places, but can be merged on its own.

The PR changes behavior in subtle ways. At the moment, the demux operator in reachability consolidates its outputs, which hides any sub-tick-interval changes. With flat containers, we cannot yet use the consolidating container builder, so we might ship more data to the next operator. The next operator happens to be an arrangement, so it'll consolidate the data on its input. This makes the situation less bad, but we might still have some longer-lived allocations.

It does not move data out of lgalloc because only rows in columnation are backed by lgalloc, but all other types use the default implementation. The spilled part in reachability is a `Vec<usize>`, so it currently can't benefit from lgalloc. This means there is no potential for a regression in this dimension.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/compute/src/logging/reachability.rs [hunk 1]
```diff
@@ -77,36 +81,44 @@ pub(super) fn construct<A: Allocate>(
         use timely::dataflow::channels::pact::Pipeline;
         let mut input = flatten.new_input(&logs, Pipeline);
 
-        let (mut updates_out, updates) = flatten.new_output::<ConsolidatingContainerBuilder<_>>();
+        type RUpdates = <(
```
Comment (frankmcsherry) [inline src/compute/src/logging/reachability.rs:84]: Maybe a comment here about what is going on? `RUpdates` isn't a clear enough name for me to unpack all the tuple components.

File: src/compute/src/logging/reachability.rs [hunk 2]
```diff
@@ -77,36 +81,44 @@ pub(super) fn construct<A: Allocate>(
         use timely::dataflow::channels::pact::Pipeline;
         let mut input = flatten.new_input(&logs, Pipeline);
 
-        let (mut updates_out, updates) = flatten.new_output::<ConsolidatingContainerBuilder<_>>();
```
Comment (antiguru) [inline src/compute/src/logging/reachability.rs:80]: I'll do some measurements to confirm that removing the eager consolidation is not a regression.

**Review by frankmcsherry — APPROVED** [review]
As far as I can tell this checks out. I do want to flag that if it is getting our toes in the water wrt how to use this stuff, it's as clear as mud at the moment. I wouldn't be able to do this myself yet. Perhaps that's fine, and the plan is to learn from this and improve, or abstract up, or something.

File: src/compute/src/logging/reachability.rs (hunk 1 above)
Comment (antiguru) [reply src/compute/src/logging/reachability.rs]: Renamed to `UpdatesRegion`.

**Review by teskje — APPROVED** [review]
LGTM.

**Pull request merged**